### PR TITLE
Doc/architecture

### PR DESCRIPTION
**Pull request by szuecs**
Skipper architecture documentation

### CODE REVIEW

**Comment by szuecs** [comment]
:+1:

File: docs/index.md [hunk 1]
```diff
@@ -1,4 +1,49 @@
 # Architecture
 
-Skipper is written as library and has also 2 binaries, named `skipper`
-and `eskip`.
+Skipper is written as library and is also a multi binary project with
```
Comment (aryszka) [inline docs/index.md]: ...as a library...

File: docs/index.md [hunk 2]
```diff
@@ -1,4 +1,49 @@
...
+information from different sources, for example static routes from an
+eskip file or dynamic routes from Kubernetes ingress objects.
+
+The `proxy` package gets the routes populated by skipper and has
+always a current routing table which will be replaced on change.
+
+A route is one entry in the routing table. A route consists of one or
+more `predicate`, that are used to find a route for a given HTTP
+request. A route can also have one or more `filter`, that can modify
+the content of the request or response.
```
Comment (aryszka) [inline docs/index.md]: I'd add here something like: A route always points to a backend, or it's a shunt, meaning that skipper serves the requests for the route, or a loopback, meaning that the request will be matched against the routing table again after filters have modified it.

File: docs/index.md [hunk 3]
```diff
@@ -1,4 +1,49 @@
...
+
+Skipper has a rich set of metrics that are exposed as json, but can be
+exported in [Prometheus](https://prometheus.io) format.
+
+![Skipper's architecture ](/img/architecture.svg)
+
+## Route processing
+
+Package `skipper` has a Go `http.Server` and does the `ListenAndServe`
+call with the `loggingHandler` wrapped `proxy`.  The `loggingHandler`
```
Comment (aryszka) [inline docs/index.md]: The `loggingHandler` is basically a middleware for the proxy providing access logs and both...

File: docs/index.md [hunk 4]
```diff
@@ -1,4 +1,49 @@
...
+proxy global ratelimits first and after that lookup the route in the
+routing table. After that skipper will apply all request filters, that
+can modify the `http.Request`. It will then check the route local
+ratelimits, the circuitbreakers and do the backend call. If the
+backend call got a TCP or TLS connection error in a loadbalanced
+route, skipper will do a retry to another backend of that loadbalanced
+group automatically. Just before the response to the caller, skipper
+will process the response filters, that can change the
+`http.Response`.
+
```
Comment (aryszka) [inline docs/index.md]: In two special cases, skipper doesn't forward the request to the backend. When the route is shunted, skipper serves the request alone, by using only the filters. When the route is a loopback, the request is passed to the routing table for finding another route, based on the changes that the filters made to the request.

**Comment by aryszka** [comment]
made some suggestions, but generally looks great!

**Comment by aryszka** [comment]
:+1:

File: docs/index.md (hunk 2 above)
Comment (szuecs) [reply docs/index.md]: I will use `<shunt>` and the same for loopback

**Comment by szuecs** [comment]
:+1:

**Comment by aryszka** [comment]
:+1:

**Pull request merged**